pdns/recursordist/rec-taskqueue.cc

```diff
@@ -22,50 +22,139 @@
 #include "rec-taskqueue.hh"
 #include "taskqueue.hh"
 #include "lock.hh"
-#include "logger.hh"
+#include "logging.hh"
 #include "stat_t.hh"
 #include "syncres.hh"
 
-static LockGuarded<pdns::TaskQueue> s_taskQueue;
-static pdns::stat_t s_almost_expired_tasks_pushed;
-static pdns::stat_t s_almost_expired_tasks_run;
-static pdns::stat_t s_almost_expired_tasks_exceptions;
+// For rate lmiting, we maintain a set of tasks recently submitted.
+class TimedSet
+{
+public:
+  TimedSet(time_t t) :
+    d_expiry_seconds(t)
+  {
+  }
+
+  uint64_t purge(time_t now)
+  {
+    // This purge is relatively sleep, as we're walking an ordered index
+    uint64_t erased = 0;
+    auto& ind = d_set.template get<time_t>();
+    auto it = ind.begin();
+    while (it != ind.end()) {
+      if (it->d_ttd < now) {
+        ++erased;
+        it = ind.erase(it);
+      }
+      else {
+        break;
+      }
+    }
+    return erased;
+  }
+
+  bool insert(time_t now, const pdns::ResolveTask& task)
+  {
+    // We periodically purge
+    if (++d_count % 1024 == 0) {
+      purge(now);
+    }
+    time_t ttd = now + d_expiry_seconds;
+    bool inserted = d_set.emplace(task, ttd).second;
+    if (!inserted) {
+      uint64_t erased = purge(now);
+      // Try again if the purge deleted at least one entry
```

Comment:
Why does a deletion allow the insert to happen?

Reply:
`d_set` can contain an equal task that is expired but not yet purged as we do only purge once every 1000 calls. So if the insert fails, we do an extra purge sweep to make sure we an expired task does not prevent the insert.